lib/internal/process/esm_loader.js

```diff
@@ -31,31 +39,91 @@ exports.importModuleDynamicallyCallback = async function(wrap, specifier) {
   throw new ERR_VM_DYNAMIC_IMPORT_CALLBACK_MISSING();
 };
 
-let ESMLoader = new Loader();
-exports.ESMLoader = ESMLoader;
-
-exports.initializeLoader = initializeLoader;
-async function initializeLoader() {
-  const { getOptionValue } = require('internal/options');
-  const userLoader = getOptionValue('--experimental-loader');
-  if (!userLoader)
-    return;
-  let cwd;
-  try {
-    cwd = process.cwd() + '/';
-  } catch {
-    cwd = 'file:///';
+let processLoader;
+exports.getLoader = () => {
+  if (!processLoader) {
+    processLoader = (async () => {
+      const loader = new Loader();
+
+      const { getOptionValue } = require('internal/options');
+      const userLoaders = getOptionValue('--experimental-loader');
+
+      if (userLoaders.length > 0) {
+        const { emitExperimentalWarning } = require('internal/util');
+        emitExperimentalWarning('--experimental-loader');
+
+        const importedLoaders = [];
+        for (let i = 0; i < userLoaders.length; i += 1) {
+          const ns = await loader.importLoader(userLoaders[i]);
+
+          if (ns.resolve !== undefined) {
+            validateFunction(ns.resolve, 'resolve');
+          }
+          if (ns.getFormat !== undefined) {
+            validateFunction(ns.getFormat, 'getFormat');
+          }
+          if (ns.getSource !== undefined) {
+            validateFunction(ns.getSource, 'getSource');
+          }
+          if (ns.transformSource !== undefined) {
+            validateFunction(ns.transformSource, 'transformSource');
+            ArrayPrototypePush(loader.transformSourceHooks, ns.transformSource);
+          }
+
+          if (ns.getGlobalPreloadCode !== undefined) {
+            validateFunction(ns.getGlobalPreloadCode, 'getGlobalPreloadCode');
+            const code = ReflectApply(ns.getGlobalPreloadCode, undefined, []);
+            validateString(code, 'getGlobalPreloadCode return value');
+            ArrayPrototypePush(loader.globalPreloadCode, code);
+          }
+
+          importedLoaders.push(ns);
+        }
+
+        let cursor = {
+          resolve: loader._resolve,
+          getFormat: loader._getFormat,
+          getSource: loader._getSource,
+        };
+        for (let i = userLoaders.length - 1; i >= 0; i -= 1) {
+          const {
+            resolve,
+            getFormat,
+            getSource,
+          } = importedLoaders[i];
+
+          const nextLoader = cursor;
+          cursor = {
+            resolve: resolve ? async (...args) => {
+              const result = await resolve(...args, nextLoader.resolve);
+              if (result === null) {
+                return nextLoader.resolve(...args);
```

Comment:
Actually, one more thing. In order to allow us to still decide what the `this` context can be (and if not doing something in this PR), can we please set the `this` value to null for these hook calls?

Reply:
it's already undefined, `nextLoader.whatever` is the async wrapper function (or the node default one)